core/src/mast/node/basic_block_node/mod.rs

```diff
@@ -384,6 +403,128 @@ impl fmt::Display for BasicBlockNodePrettyPrint<'_> {
     }
 }
 
+// DECORATOR ITERATION
+// ================================================================================================
+
+/// Iterator used to iterate through the decorator list of a span block
+/// while executing operation batches of a span block.
+///
+/// This lets the caller iterate through a Decorator list with indexes that match the
+/// standard (padded) representation of a basic block.
+pub struct DecoratorOpLinkIterator<'a> {
+    before: core::slice::Iter<'a, DecoratorId>,
+    middle: core::slice::Iter<'a, (usize, DecoratorId)>,
+    after: core::slice::Iter<'a, DecoratorId>,
+    total_ops: usize,
+    seg: Segment,
+}
+
+// Driver of the Iterators' state machine
+enum Segment {
+    Before,
+    Middle,
+    After,
+    Done,
+}
+
+impl<'a> DecoratorOpLinkIterator<'a> {
+    pub fn new(
+        before_enter: &'a [DecoratorId],
+        decorators: &'a DecoratorList,
+        after_exit: &'a [DecoratorId],
+        total_operations: usize,
+    ) -> Self {
+        Self {
+            before: before_enter.iter(),
+            middle: decorators.iter(),
+            after: after_exit.iter(),
+            total_ops: total_operations,
+            seg: Segment::Before,
+        }
+    }
+
+    /// Optional: yield only if the next item corresponds to the given op index.
+    /// - before_enter items map to op 0
+    /// - middle items use their stored position
+    /// - after_exit items map to `total_ops`
+    //
+    // some decorators are pegged on an operation index equal to the total number of
+    // operations since decorators are meant to be executed before the operation
+    // they are attached to, this allows them to be executed after the last
+    // operation has been executed.
+    #[inline]
+    pub fn next_filtered(&mut self, pos: usize) -> Option<(usize, DecoratorId)> {
+        let should_yield: bool;
+        'segwalk: loop {
+            match self.seg {
+                Segment::Before => {
+                    if self.before.clone().next().is_some() {
+                        should_yield = pos == 0;
+                        break 'segwalk;
+                    }
+                    self.seg = Segment::Middle;
+                },
+                Segment::Middle => {
+                    if let Some(&(p, _)) = self.middle.clone().next() {
+                        should_yield = pos == p;
+                        break 'segwalk;
+                    }
+                    self.seg = Segment::After;
+                },
+                Segment::After => {
+                    if self.after.clone().next().is_some() {
```

Comment:
While the iterator should be cheap to clone, we can use `peekable()`.

```suggestion
                    if self.after.peekable().next().is_some() {
```